warehouse/query-core/src/main/java/datawave/query/lucene/visitors/PrintingVisitor.java

```diff
@@ -0,0 +1,325 @@
+package datawave.query.lucene.visitors;
+
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.hadoop.util.Lists;
+import org.apache.lucene.queryparser.flexible.core.nodes.AndQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.AnyQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.BooleanQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.BoostQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.DeletedQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.FieldQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.FunctionQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.FuzzyQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.GroupQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.MatchAllDocsQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.MatchNoDocsQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.ModifierQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.NoTokenFoundQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.NotBooleanQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.OpaqueQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.OrQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.PathQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.PhraseSlopQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.ProximityQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.QueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.QuotedFieldQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.SlopQueryNode;
+import org.apache.lucene.queryparser.flexible.core.nodes.TokenizedPhraseQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.AbstractRangeQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.BooleanModifierNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.MultiPhraseQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.PointQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.PointRangeQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.PrefixWildcardQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.RegexpQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.SynonymQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.TermRangeQueryNode;
+import org.apache.lucene.queryparser.flexible.standard.nodes.WildcardQueryNode;
+
+import datawave.query.Constants;
+
+public class PrintingVisitor extends BaseVisitor {
```

Comment:
I am not seeing this class being used anywhere, what am I missing.  Also what functionality does this provide that the LuceneQueryStringBuildingVisitor does not already supply?

Reply:
This class is analogous to the PrintingVisitor for JEXL queries. It prints the overall structure of LUCENE query trees so that we can the hierarchy of nodes and their details. We used it extensively when examining trees while writing tests. It is currently now used in at least one place to log the LUCENE query after the initial parse in `ShardQueryLogic.validateQuery().`

The LuceneQueryStringBuildingVisitor is analogous to JexlStringBuildingVisitor. It returns a LUCENE query tree as a query string, but does not show the actual structure of the tree.